File: scripts/block/effect/LiquidUnloader.js
/**
 * @author guiY<guiYMOUR>
 * @Extra mod <https://github.com/guiYMOUR/mindustry-Extra-Utilities-mod>
 */
//引用部分，类似import，对应的是exports导出
const items = require("game/items");

//速度，一般性不用改
const speed = 2;

//液体卸货器就是用卸货器的码改液体源
const lu = extendContent(LiquidSource, "liquid-unloader", {
    drawRequestConfig(req, list){
        this.drawRequestConfigCenter(req, req.config, Core.atlas.find("btm-liquid-unloader-centre"), true);
    },
});
lu.buildType = prov(() => {
    var dumpingTo = null;
    var offset = 0;
    var liquidBegin = null;
    
    const block = lu;
    var source = null;
    var proximity = null;
    
    return new JavaAdapter(LiquidSource.LiquidSourceBuild, {
        updateTile(){
            source = this.source;
            proximity = this.proximity;
            if(liquidBegin != source){
                this.liquids.clear();
                liquidBegin = source;
            }
            for(var i = 0; i < proximity.size; i++){
                var pos = (offset + i) % proximity.size;
                var other = proximity.get(pos);

                if(other.interactable(this.team) && other.block.hasLiquids && !(other instanceof LiquidBlock.LiquidBuild && other.block.size == 1) && source != null && other.liquids.get(source) > 0){
                    dumpingTo = other;
                    if(this.liquids.total() < block.liquidCapacity){
                        var amount = Math.min(speed, other.liquids.get(source));
                        this.liquids.add(source, amount);
                        other.liquids.remove(source, amount);
                    }
                }
            }
            if(proximity.size > 0){
                offset ++;
                offset %= proximity.size;
            }
            this.dumpLiquid(this.liquids.current());
        },
        canDumpLiquid(to, liquid){
            return to != dumpingTo;
        },
        draw(){
            Draw.rect(Core.atlas.find("btm-liquid-unloader"), this.x, this.y);
            if(this.source == null){
                Draw.rect("cross", this.x, this.y);
            }else{
                Draw.color(this.source.color);
                Draw.rect(Core.atlas.find("btm-liquid-unloader-centre"), this.x, this.y);
                Draw.color();
            }
        },
    }, lu);
});
lu.health = 70;
lu.liquidCapacity = 10;
lu.requirements = ItemStack.with(
    Items.metaglass, 18,
    Items.silicon, 18,
    Items.titanium, 12
);
lu.buildVisibility = BuildVisibility.shown;
lu.category = Category.effect;
exports.lu = lu;

File: src/ExtraUtilities/worlds/blocks/liquid/LiquidUnloader.java
package ExtraUtilities.worlds.blocks.liquid;

import arc.Core;
import arc.graphics.Color;
import arc.graphics.g2d.*;
import arc.scene.ui.layout.Table;
import arc.util.Eachable;
import arc.util.io.*;
import mindustry.entities.units.BuildPlan;
import mindustry.gen.Building;
import mindustry.type.Liquid;
import mindustry.world.Block;
import mindustry.world.blocks.ItemSelection;
import mindustry.world.blocks.liquid.LiquidBlock.*;
import mindustry.world.meta.*;

import static mindustry.Vars.*;

public class LiquidUnloader extends Block {
    public String center;
    public float speed = 3f;

    public LiquidUnloader(String name){
        super(name);

        update = true;
        solid = true;
        hasLiquids = true;
        liquidCapacity = 10f;
        configurable = true;
        outputsLiquid = true;
        saveConfig = true;
        noUpdateDisabled = true;
        displayFlow = false;
        group = BlockGroup.liquids;
        envEnabled = Env.any;
        clearOnDoubleTap = true;

        config(Liquid.class, (LiquidUnloaderBuild tile, Liquid l) -> tile.sortLiquid = l);
        configClear((LiquidUnloaderBuild tile) -> tile.sortLiquid = null);
    }

    @Override
    public void load() {
        super.load();
        center = name + "-center";
    }

    @Override
    public void setBars(){
        super.setBars();

        removeBar("liquid");
    }

    @Override
    public void drawPlanConfig(BuildPlan plan, Eachable<BuildPlan> list){
        drawPlanConfigCenter(plan, plan.config, center, true);
    }


    public class LiquidUnloaderBuild extends Building {
        public Liquid sortLiquid = null;
        public Liquid lastSort = null;
        public Building dumpingTo = null;
        public float offset = 0f;

        @Override
        public void updateTile() {
            if(lastSort != sortLiquid){
                liquids.clear();
                lastSort = sortLiquid;
            }
            for(int i = 0; i < proximity.size; i++){
                int pos = (int)(offset + i) % proximity.size;
                Building other = proximity.get(pos);

                if(other.interactable(team) && other.block.hasLiquids && !(other instanceof LiquidBuild && other.block.size == 1) && sortLiquid != null && other.liquids.get(sortLiquid) > 0){
                    dumpingTo = other;
                    if(liquids.get(sortLiquid) < block.liquidCapacity){
                        float amount = Math.min(speed, other.liquids.get(sortLiquid));
                        liquids.add(sortLiquid, amount);
                        other.liquids.remove(sortLiquid, amount);
                    }
                }
            }
            if(proximity.size > 0){
                offset ++;
                offset %= proximity.size;
            }
            this.dumpLiquid(liquids.current());
        }

        @Override
        public boolean canDumpLiquid(Building to, Liquid liquid) {
            return to != dumpingTo;
        }

        @Override
        public void draw() {
            super.draw();
            Draw.color(sortLiquid == null ? Color.clear : sortLiquid.color);
            Draw.rect(Core.atlas.find(center), x, y);
            Draw.color();
        }

        @Override
        public void buildConfiguration(Table table){
            ItemSelection.buildTable(LiquidUnloader.this, table, content.liquids(), () -> sortLiquid, this::configure);
        }

        @Override
        public Liquid config(){
            return sortLiquid;
        }

        @Override
        public byte version(){
            return 1;
        }

        @Override
        public void write(Writes write){
            super.write(write);
            write.s(sortLiquid == null ? -1 : sortLiquid.id);
        }

        @Override
        public void read(Reads read, byte revision){
            super.read(read, revision);
            int id = revision == 1 ? read.s() : read.b();
            sortLiquid = id == -1 ? null : content.liquid(id);
        }
    }
}